Utils and libraries: Drops the boost/algorithm/string/split.hpp dependency

l2a5b1 opened the pull request:
This pull request drops the `boost/algorithm/string/split.hpp` dependency from the project. It replaces `boost::split` with a custom template function `Split` that has a similar API and returns exactly the same results as `boost::split`.

In addition this pull request refactors an instance of `boost::algorithm::trim_right`, which was implicitly made available via the `boost/algorithm/string.hpp` dependency to prevent having to introduce a new dependency `boost/algorithm/string/trim.hpp`.

The `#include <boost/algorithm/string/predicate.hpp>` in `src/wallet/rpcdump.cpp` will be addressed in #13656 if that PR is merged after this one; or in this PR should #13656 be merged first.

The test vectors in the accompanying unit test have been validated against `boost::split`.

MarcoFalke commented:
Hmm, this is just replacing a boost header with a self-written header that offers the same functionality and comes with unit tests. Do we really want to re-implement all of boost we are currently using? I mean straightforward replacing `boost::` with `std::` is fine, but re-implementing boost should be one of our low priority goals, no?

251Labs commented:
That's a fair point @MarcoFalke 👍 

The goal was not to re-implement boost. This is an attempt to drop the `boost::split` dependency with as little changes to the existing codebase as possible.

I have eliminated boost where it was trivial to do: `boost::algorithm::is_any_of` has been refactored to a `std::string` and `boost::algorithm::trim_right` to a `std::string:erase`.

There are many ways to simplify this function:  e.g. no support for both `std::vector` and `std::set`; no support for `boost::token_compress_on`.

I didn't want to go there initially, because I didn't want to break things and ease review by providing something that is straightforward to validate. Right now it is easy to swap the `boost::split` and `Split` functions in the unit test to validate if the external behaviour of `Split` still works as expected.

That said, I don't have strong opinion on this. I am happy to make changes where needed; or close the pull request if this is not what you're looking for.

I just hope I haven't messed up 😅

practicalswift commented:
Concept ACK

The short self-contained non-Boost version is much easier to reason about. Let's get rid of Boost! :-)

MarcoFalke commented:
@practicalswift Sure, but it has to go through review to verify that it supports everything that was supported by boost the way we used it.

We are stuck with boost until we adopt C++17, since implementing `std::optional` or `std::variant` on our own wouldn't be too wise, imo. So at least there is no rush in getting rid of non-trivial boost features that are not already in `std::`.

practicalswift commented:
@MarcoFalke Agreed – no rush in removing non-trivial Boost features. Let’s do it incrementally starting with the low hanging fruit :-)

251Labs commented:
> So at least there is no rush in getting rid of non-trivial boost features that are not already in `std::`

@MarcoFalke, apologies about this. I have totally misinterpreted the objective of the Boost migration.

MarcoFalke commented:
Sorry for being so pessimistic about it. You can definitely leave this open and see what other reviewers think. I was just saying that we have about 3 years time to finish up everything.

251Labs commented:
Thanks @MarcoFalke!

practicalswift commented:
@251Labs Text splitting should be in the low hanging fruit category I mentioned :-)

sipa commented:
Note there is a function with similar functionality in #13697, but operating on `Span<const char>`s rather than `std::string`s.

practicalswift commented:
@251Labs See if you can base your work in this PR on @sipa:s code in #13697 :-)

Empact commented on src/utilsplitstring.h:
```diff
@@ -0,0 +1,60 @@
+
+    for (auto it = begin; it < end;) {
+        bool foundSeparator = false;
+        auto tokenIt = it;
```
nit: this would be a bit simpler in implementation if it always referred to the end of the token, and you stashed the beginning here and continued to operate on it. Does that make sense?

```C++
for (auto it = begin; it < end; ++it) {
        bool foundSeparator = false;
        auto start = it;
        while (it < end && !(foundSeparator = anyOfSeparator.find(*it) != std::string::npos)) {
            ++it;
        }
        if (it != begin && (!mergeEmpty || start != it)) {
            *insertIt = std::string(start, it);
        }
        if (foundSeparator && (it == begin || it == end - 1)) {
            *insertIt = "";
        }
    }
```

251Labs replied:
Thanks for the review @Empact!

I have updated the PR and addressed your feedback.

Please note that the separate `if` statements below are intentional and cover the case where `it` equals `begin` and `it` equals `end - 1`, in which case two empty tokens need to be inserted to be boost compliant.

```c++
        if (it == begin) {
            *insertIt = "";
        }
        if (it == end - 1) {
            *insertIt = "";
        }
```
This behavior is validated by the following unit test in `utilsplitstring_tests.cpp`:
```c++
       Split(result, ",", ",", true);
       std::vector<std::string> expected = {"", ""};
```

Empact commented on src/utilsplitstring.h:
```diff
@@ -0,0 +1,60 @@
+    static_assert(
+        std::is_same<std::vector<std::string>, ContainerT>::value ||
+        std::is_same<std::set<std::string>, ContainerT>::value,
+        "ContainerT must be of type std::vector<std::string> or std::set<std::string>");
```
I'm not sure this restriction is helpful - absent it the args would be limited to whatever is compatible with `std::inserter`, which could be fine...

251Labs replied:
I did add the `static_assert`, because there were no other use cases in the codebase and didn't want to make any assumptions for containers other than `std::vector` and `std::set`. That said, you've raised a valid point and I am happy to remove it.

251Labs commented:
5502212 addresses review comments https://github.com/bitcoin/bitcoin/pull/13751#discussion_r207340164 and https://github.com/bitcoin/bitcoin/pull/13751#discussion_r207341411

laanwj commented:
#13697 was merged
we now have in `src/script/descriptor.cpp`

`std::vector<Span<const char>> Split(const Span<const char>& sp, char sep)`

would make sense to move that to a public header instead and use that

251Labs commented:
Hey, thanks @laanwj!

Yeah, good point. 

I had been thinking about this a while back when @sipa and @practicalswift hinted at the same, but didn't see an easy non-invasive way to do this, because the code that uses `boost::split` relies on specific external behaviour that`std::vector<Span<const char>> Split(const Span<const char>& sp, char sep)` doesn't provide.

For example:
1. `torcontroller.cpp` expects the result in a `std::set<std::string>` instead of `std::vector<std::string>`;
2. `httprpc.cpp` depends on `boost::split`'s capability to split on different separators;
3. `core_read.cpp` depends on `boost::split`'s capability to merge adjacent separators

Right now, the refactor seems trivial enough, because only the `boost::split` calls had to be replaced with calls to a custom function with identical external behaviour (as asserted by the unit tests, which can also be ran with the original `boost::split` function). No other code had to be refactored. 

Dropping split from this pull request and trying to make the codebase work with `std::vector<Span<const char>> Split(const Span<const char>& sp, char sep)` seems very invasive, but maybe I am missing something obvious.

Alternatively, one split function could use the other one, e.g. `std::vector<Span<const char>> Split(const Span<const char>& sp, char sep)` from `descriptor` could use `template<typename ContainerT>void Split(ContainerT& tokens, const std::string& str, const std::string& anyOfSeparator, bool mergeEmpty = false)` from this pull request.

I am not sure about the options here. Thoughts?

practicalswift commented on src/utilsplitstring.h:
```diff
@@ -0,0 +1,59 @@
+    for (auto it = begin; it < end; ++it) {
+        bool foundSeparator = false;
+        auto start = it;
+        while (it < end && !(foundSeparator = anyOfSeparator.find(*it) != std::string::npos)) {
```
Try to reformulate without mixing assignment and comparison. That would make it easier to read and reason about :-)

251Labs replied:
Hey, thanks @practicalswift! Yes, absolutely. Please see 3704125.

practicalswift commented on src/utilsplitstring.h:
```diff
@@ -0,0 +1,63 @@
+    const auto end = str.cend();
+
+    for (auto it = begin; it < end; ++it) {
+        bool foundSeparator;
```
Initialize to `false` :-)

laanwj commented:
@251Labs fair enough, if it's so much hassle it is probably better to add another function, and leave the other split() private to the descriptor parsing code

251Labs commented on src/utilsplitstring.h:
```diff
@@ -0,0 +1,63 @@
+            *insertIt = std::string(start, it);
+        }
+
+        if (foundSeparator) {
```
Hey @practicalswift, thanks! I don't think `foundSeparator` can be uninitialized at this point, because it will always go at least one time through the inner for loop where it is initialized, but I will revert back to explicit initialization of this variable for clarity.

251Labs commented:
Rebased 5005a8f

251Labs commented on src/util/splitstring.h:
```diff
@@ -0,0 +1,63 @@
+ * @param[in] mergeEmpty        Set to true to merge adjacent separators (empty tokens); otherwise false (default).
+ */
+template<typename ContainerT>
+void Split(ContainerT& tokens, const std::string& str, const std::string& anyOfSeparator, bool mergeEmpty = false)
```
Thanks @sipa, sure! See e236c47.

251Labs commented:
e236c47 renames lower camel case variable names to snake_case equivalents.

251Labs commented:
rebased ab2a18d

practicalswift commented:
@251Labs Looks good. I’m near a utACK - could you rebase on top of #15043 and add a fuzzer for the split method which would split on random input and make sure the result matches the result from `boost::split`?

luke-jr reviewed (changes requested):
NACK. Using dependencies is strictly better than reinventing stuff for no reason.

Empact commented:
@luke-jr I thought we've been working toward a long term goal of removing boost entirely. Is that not right?

practicalswift commented:
In contrast to @luke-jr: let me re-iterate my concept ACK. IMO we should take small incremental steps towards a code base free from Boost which is exactly what this PR does.

tiendq commented:
@251Labs @MarcoFalke I just came here from a Boost to C++ 11 migration project page. Where can I find more information of migration plan? (if there is one). Just curious to know in case I could give it a hand :)

MarcoFalke commented:
The migration is pretty much just to replace `boost::` with `std::`. Though, the low hanging fruits in C++11 have already been collected (see https://github.com/bitcoin/bitcoin/projects/3).

MarcoFalke commented:
If `boost::split` was the last dependency we used from boost, this pull could be merged, but I don't see that coming any time soon.

251Labs commented:
rebased d7ff8eb

251Labs commented:
rebased 388257f

251Labs commented:
rebased 9f811c3

251Labs commented:
rebased 402da1c

Travis build error is unrelated:
```
Error! Initial build successful, but not enough time remains to run later build stages and tests. 
Please manually re-run this job by using the travis restart button or asking a bitcoin maintainer 
to restart. The next run should not time out because the build cache has been saved.
```

promag commented:
@251Labs restarted.

251Labs commented:
Thanks @promag!

251Labs commented:
Rebased 6186091

l2a5b1 commented:
rebased 47a758b

l2a5b1 commented:
Rebased 5730465

l2a5b1 committed:
Drops boost/algorithm/string/split.hpp

This commit drops the `boost/algorithm/string/split.hpp` dependency from
the project.

It replaces `boost::split` with a custom function `Split` that has
an identical API and returns exactly the same results as `boost::split`
to ease refactoring.

In addition this commit refactors an instance of `boost::algorithm::trim_right`
which was implicitly made available via the `boost/algorithm/string.hpp`
dependency to prevent having to introduce a new dependency
`boost/algorithm/string/trim.hpp`.

fanquake commented:
Thanks for the work you've done here. However the project has decided that while it remains impossible to remove all of our Boost usage, we are no longer going to remove smaller components by rewriting them ourselves.

l2a5b1 commented:
Thanks @fanquake.

Context FWIW:

[bitcoin-core-dev/2019-10-31.log](http://gnusha.org/bitcoin-core-dev/2019-10-31.log)

> 12:08 < wumpus> #topic close Boost -> C++11 migration project for now (wumpus)
> 12:09 < warren> too much change required?
> 12:09 < wumpus> so it looks like all the low-hanging fruit for replacing boost has been picked now
> 12:10 < wumpus> what remains is hard to replace with C++11 (results in very verbose code, locale dependent legacy C, or introduces harder to maintain platform-specific code)
> 12:10 < jeremyrubin> It's basically just multiindex and boost::thread now right?
> 12:10 < wumpus> it's kind of a time wasting trap for developers now (regarding PRs like #17245)
> 12:10 < gribble> https://github.com/bitcoin/bitcoin/issues/17245 | wallet: Remove Boost from DecodeDumpTime by elichai . Pull Request #17245 . bitcoin/bitcoin . GitHub
> 12:10 < wumpus> nah, people stumble over the sleep and date/time handling stuff every time
> 12:11 < wumpus> #17307 might still be worthwhile
> 12:11 < gribble> https://github.com/bitcoin/bitcoin/issues/17307 | Stop using `boost::thread_group` . Issue #17307 . bitcoin/bitcoin . GitHub
> 12:11 < jeremyrubin> What about just checking in those specific copies of boost library code
> 12:11 < sipa> after 17307, won't removing boost::threa dbe a lot easier?
> 12:11 < jeremyrubin> Or are they too large/dependent on the rest of boost types
> 12:11 < wumpus> but it needs to be focused, we don't want to close the same PRs time after time that don't really make it
> 12:12 < wumpus> I think having that project open sends the wrong messag
> 12:12 < wumpus> we can't replace boost right now
> 12:12 < jeremyrubin> 17307, having worked on replacing thread_group in the checkqueue before, scares me a bit
> 12:12 < sipa> agree there
> 12:12 < wumpus> there's no need to replace, say, small string handling functions with our own impelentation, before we have a vision to replace the rest
> 12:13 < sipa> right; until we have a reasonable way to remove multiindex (which i'm not sure will ever happen), getting rid of boost entirely is not really useful
> 12:13 < dongcarl> Just so it's out there... we can maybe run bcp at some point when there's only one or two things left https://www.boost.org/doc/libs/1_71_0/tools/bcp/doc/html/index.html
> 12:13 < jeremyrubin> dongcarl: ++
> 12:13 < sipa> i do think there are individual improvements possible that dkn't go all the way, like removing boost::thread
> 12:14 < jnewbery> I agree that if it's not a priority then the project should be closed. It can always be re-opened later if necessary.
> 12:14 < wumpus> but in any case it doesn't seem like it needs a big coordinated project anymore
> 12:14 < sipa> how many non-headers-only boost libs are we still using?
> 12:14 < sipa> wumpus: agree
> 12:14 < jnewbery> (leaving a comment in the project description explaining why it's been closed)
> 12:14 < wumpus> when C++17 is allowed, it can be reopened
> 12:15 < fanquake> Guess #13751 can be closed with the same rationale as well then.
> 12:15 < gribble> https://github.com/bitcoin/bitcoin/issues/13751 | Utils and libraries: Drops the boost/algorithm/string/split.hpp dependency by l2a5b1 . Pull Request #13751 . bitcoin/bitcoin . GitHub
> 12:15 < wumpus> fanquake: yes
